src/api/common/connection.js

```diff
@@ -0,0 +1,192 @@
+         type === 'transaction' ||
+         type === 'path_find';
+}
+
+class RippledError extends Error {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+    this.message = message;
+    Error.captureStackTrace(this, this.constructor.name);
+  }
+}
+
+class ConnectionError extends Error {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+    this.message = message;
+    Error.captureStackTrace(this, this.constructor.name);
+  }
+}
+
+class DisconnectedError extends ConnectionError {
+  constructor(message) {
+    super(message);
+  }
+}
+
+class TimeoutError extends ConnectionError {
+  constructor(message) {
+    super(message);
+  }
+}
+
+class UnexpectedError extends ConnectionError {
+  constructor(message) {
+    super(message);
+  }
+}
+
+class Connection extends EventEmitter {
+  constructor(url, options = {}) {
+    super();
+    this._url = url;
+    this._timeout = options.timeout || (20 * 1000);
+    this._ws = null;
+    this._nextRequestID = 1;
+  }
+
+  _onMessage(message) {
+    try {
+      const data = JSON.parse(message);
+      if (data.type === 'response') {
+        if (!(Number.isInteger(data.id) && data.id >= 0)) {
+          throw new UnexpectedError('valid id not found in response');
+        }
+        this.emit(data.id.toString(), data);
+      } else if (isStreamMessageType(data.type)) {
+        this.emit(data.type, data);
+      } else if (data.type === undefined && data.error) {
+        this.emit('error', data.error, data.error_message);  // e.g. slowDown
+      } else {
+        throw new UnexpectedError('unrecognized message type: ' + data.type);
+      }
+    } catch (error) {
+      this.emit('error', 'badMessage', message);
+    }
+  }
+
+  get state() {
+    return this._ws ? this._ws.readyState : WebSocket.CLOSED;
+  }
+
+  _onUnexpectedClose() {
+    this.connect().then();
+  }
+
+  connect() {
+    return new Promise((resolve) => {
+      if (this.state === WebSocket.OPEN) {
+        resolve();
+      } else if (this.state === WebSocket.CONNECTING) {
+        this._ws.once('open', resolve);
+      } else {
+        this._ws = new WebSocket(this._url);
+        this._ws.on('message', this._onMessage.bind(this));
+        this._ws.once('close', () => this._onUnexpectedClose);
+        this._ws.once('open', resolve);
+      }
+    });
+  }
+
+  disconnect() {
+    return new Promise((resolve) => {
+      if (this.state === WebSocket.CLOSED) {
+        resolve();
+      } else if (this.state === WebSocket.CLOSING) {
+        this._ws.once('close', resolve);
+      } else {
+        this._ws.removeListener('close', this._onUnexpectedClose);
+        this._ws.once('close', resolve);
+        this._ws.close();
+      }
+    });
+  }
+
+  reconnect() {
+    return this.disconnect().then(() => this.connect());
+  }
+
+  _send(message) {
+    return new Promise((resolve, reject) => {
+      try {
+        this._ws.send(message);
```

Comment:
At least in `ws` module, errors are thrown as exceptions if there is no callback - I've hit that when I've worked on that `slowDown` error